File: App/CommissionersCartel/Features/Members/MembersViewModel.swift
import Foundation
import Observation
import CartelCore

@Observable
@MainActor
final class MembersViewModel {
    /// A manager paired with the team they own. ESPN keeps these in separate
    /// arrays linked by SWID, so the join happens once here.
    struct Entry: Identifiable {
        let manager: Manager
        let team: Team?
        /// Nil when Supabase has no bio for this team, or is unreachable.
        var bio: TeamBio?
        /// Where the league's published power ranking puts them. Preferred
        /// over ESPN's playoff seed, which before a ball is kicked is not a
        /// ranking at all — it had the commissioner twelfth while the power
        /// ranking had him fourth, which reads as one of them being broken.
        var powerRank: Int?

        var id: String { manager.id }
    }

    /// Entries grouped under a division heading. A league without divisions
    /// produces a single unnamed group, so the view renders one flat list
    /// without needing a separate code path.
    struct Group: Identifiable {
        let division: Division?
        let entries: [Entry]

        var id: Int { division?.id ?? -1 }
        var title: String? { division?.name }
    }

    private(set) var state: Loadable<[Group]> = .idle

    /// Pull-to-refresh. Drops the ESPN cache first, so this actually goes to
    /// the network rather than replaying a response up to the cache TTL old.
    func refresh(using environment: AppEnvironment) async {
        await environment.leagueData.refresh()
        await load(using: environment, showSpinner: false)
    }

    func load(using environment: AppEnvironment, showSpinner: Bool = true) async {
        if showSpinner, state.isInitialLoad { state = .loading }

        let leagueData = environment.leagueData
        let content = environment.content
        let result = await loadState { () -> [Group] in
            async let league = leagueData.league()
            async let managers = leagueData.managers()
            async let teams = leagueData.teams()
            // Flavour text is a nice-to-have. A Supabase outage should cost us
            // the bios, not the whole members list.
            async let bios = try? await content.teamBios(season: Season.current())
            async let rankings = try? await content.powerRankings(season: Season.current())

            let allTeams = try await teams
            // One lookup per owner id, since a team can be co-owned.
            var teamByOwner: [String: Team] = [:]
            for team in allTeams {
                for owner in team.ownerIDs where teamByOwner[owner] == nil {
                    teamByOwner[owner] = team
                }
            }

            let bioByTeam = await bios ?? [:]
            let rankByTeam = Dictionary(
                uniqueKeysWithValues: (await rankings ?? []).map { ($0.teamID, $0.rank) })
            // Hand the widget the one thing it cannot work out for itself.
            // Done here rather than at claim time so it self-heals: any launch
            // that loads the roster refreshes it.
            await Self.shareClaimedTeam(teams: allTeams, managers: try await managers)
            let entries = try await managers
                .map { manager -> Entry in
                    let team = teamByOwner[manager.id]
                    return Entry(
                        manager: manager,
                        team: team,
                        bio: team.flatMap { bioByTeam[$0.id] },
                        powerRank: team.flatMap { rankByTeam[$0.id] }
                    )
                }
                // Standings order where ESPN gives us one. Before the season
                // starts every seed is absent, so fall back to record and then
                // to name — otherwise the list order is whatever ESPN felt like.
                .sorted { lhs, rhs in
                    // The published ranking first, so the number shown and the
                    // order agree with each other.
                    if let left = lhs.powerRank, let right = rhs.powerRank {
                        return left < right
                    }
                    switch (lhs.team?.playoffSeed, rhs.team?.playoffSeed) {
                    case let (l?, r?):
                        return l < r
                    case (nil, _?):
                        return false
                    case (_?, nil):
                        return true
                    case (nil, nil):
                        let left = lhs.team?.record, right = rhs.team?.record
                        let leftPct = left?.winPercentage ?? -1
                        let rightPct = right?.winPercentage ?? -1
                        if leftPct != rightPct { return leftPct > rightPct }
                        let leftFor = left?.pointsFor ?? 0
                        let rightFor = right?.pointsFor ?? 0
                        if leftFor != rightFor { return leftFor > rightFor }
                        return lhs.manager.fullName < rhs.manager.fullName
                    }
                }

            let divisions = try await league.divisions
            guard divisions.count > 1 else {
                return [Group(division: nil, entries: entries)]
            }

            // Managers whose team ESPN did not place in a division would
            // otherwise vanish from a grouped list entirely.
            var grouped = divisions.map { division in
                Group(
                    division: division,
                    entries: entries.filter { $0.team?.divisionID == division.id }
                )
            }
            let knownIDs = Set(divisions.map(\.id))
            let ungrouped = entries.filter { entry in
                guard let id = entry.team?.divisionID else { return true }
                return !knownIDs.contains(id)
            }
            if !ungrouped.isEmpty {
                grouped.append(Group(division: nil, entries: ungrouped))
            }
            return grouped.filter { !$0.entries.isEmpty }
        }

        if let result { state = result }
    }

    /// Records which team belongs to the signed-in member, for the widget.
    ///
    /// The widget runs in its own process with no session, so it cannot ask.
    /// It reads this from the shared app group and fetches the score itself.
    private static func shareClaimedTeam(teams: [Team], managers: [Manager]) async {
        guard let swid = KeychainStore.string(for: .espnSWID)?
            .trimmingCharacters(in: .whitespacesAndNewlines), !swid.isEmpty
        else { return }

        let normalised = swid.uppercased()
        guard let team = teams.first(where: { team in
            team.ownerIDs.contains { $0.uppercased() == normalised }
        }) else { return }

        SharedStore.claimedTeamID = team.id
        SharedStore.claimedTeamName = team.name
    }

    /// Flattened, for the navigation destination lookup.
    var allEntries: [Entry] {
        state.value?.flatMap(\.entries) ?? []
    }
}

File: android/app/src/main/java/com/commissionerscartel/app/feature/members/MembersViewModel.kt
package com.commissionerscartel.app.feature.members

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.commissionerscartel.app.data.AppGraph
import com.commissionerscartel.app.data.Division
import com.commissionerscartel.app.data.EspnMapper
import com.commissionerscartel.app.data.Manager
import com.commissionerscartel.app.data.Session
import com.commissionerscartel.app.data.Supabase
import com.commissionerscartel.app.widget.WidgetStore
import com.commissionerscartel.app.data.Team
import com.commissionerscartel.app.data.TeamBio
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/** A manager paired with their team and bio. ESPN keeps these apart. */
data class MemberEntry(
    val manager: Manager,
    val team: Team?,
    val bio: TeamBio?,
)

data class MemberGroup(val division: Division?, val entries: List<MemberEntry>)

sealed interface MembersState {
    data object Loading : MembersState
    data class Loaded(val groups: List<MemberGroup>) : MembersState
    data class Failed(val message: String) : MembersState
}

class MembersViewModel(application: Application) : AndroidViewModel(application) {
    private val _state = MutableStateFlow<MembersState>(MembersState.Loading)
    val state: StateFlow<MembersState> = _state.asStateFlow()

    init { load() }

    fun refresh() = load(force = true)

    private fun load(force: Boolean = false) {
        viewModelScope.launch {
            _state.value = runCatching {
                val payload = if (force) AppGraph.espn.refresh() else AppGraph.espn.payload()
                val league = EspnMapper.league(payload, AppGraph.season)
                val teams = EspnMapper.teams(payload, AppGraph.espn)
                val managers = EspnMapper.managers(payload)

                // Flavour text is a nice-to-have; a Supabase outage should cost
                // the bios, not the roster.
                val bios = runCatching { Supabase.teamBios(league.season) }.getOrDefault(emptyMap())

                // One lookup per owner, since a team can be co-owned.
                val teamByOwner = buildMap {
                    teams.forEach { team -> team.ownerIds.forEach { putIfAbsent(it, team) } }
                }

                // Hand the widget the one thing it cannot work out for itself.
                // Done on every roster load so it self-heals rather than only
                // being set at claim time.
                shareClaimedTeam(teams)

                val entries = managers
                    .map { manager ->
                        val team = teamByOwner[manager.id]
                        MemberEntry(manager, team, team?.let { bios[it.id] })
                    }
                    .sortedWith(entryOrder)

                group(entries, league.divisions)
            }.fold(
                onSuccess = { MembersState.Loaded(it) },
                onFailure = { MembersState.Failed(it.message ?: "Couldn't load the roster.") },
            )
        }
    }

    /**
     * Records which team belongs to this member, for the home screen widget.
     *
     * The widget has no session, so it cannot ask. It reads this and fetches
     * the score itself.
     */
    private suspend fun shareClaimedTeam(teams: List<Team>) {
        val swid = runCatching { Session.claimedTeamSwid() }.getOrNull()?.trim()
        if (swid.isNullOrEmpty()) return
        val mine = teams.firstOrNull { team ->
            team.ownerIds.any { it.equals(swid, ignoreCase = true) }
        } ?: return
        WidgetStore.setClaimedTeam(getApplication(), mine.id, mine.name)
    }

    /**
     * Standings order where ESPN gives one. Every seed is absent before the
     * season starts, so fall back to record and then name — otherwise the list
     * order is whatever ESPN felt like that request.
     */
    private val entryOrder = Comparator<MemberEntry> { a, b ->
        val seedA = a.team?.playoffSeed
        val seedB = b.team?.playoffSeed
        when {
            seedA != null && seedB != null -> seedA.compareTo(seedB)
            seedA != null -> -1
            seedB != null -> 1
            else -> {
                val pctA = a.team?.record?.winPercentage ?: -1.0
                val pctB = b.team?.record?.winPercentage ?: -1.0
                if (pctA != pctB) pctB.compareTo(pctA)
                else {
                    val forA = a.team?.record?.pointsFor ?: 0.0
                    val forB = b.team?.record?.pointsFor ?: 0.0
                    if (forA != forB) forB.compareTo(forA)
                    else a.manager.fullName.compareTo(b.manager.fullName)
                }
            }
        }
    }

    private fun group(entries: List<MemberEntry>, divisions: List<Division>): List<MemberGroup> {
        if (divisions.size <= 1) return listOf(MemberGroup(null, entries))

        val grouped = divisions.map { division ->
            MemberGroup(division, entries.filter { it.team?.divisionId == division.id })
        }.toMutableList()

        // A manager whose team ESPN did not place in a division would otherwise
        // vanish from the list entirely.
        val known = divisions.map { it.id }.toSet()
        val ungrouped = entries.filter { it.team?.divisionId?.let { id -> id !in known } ?: true }
        if (ungrouped.isNotEmpty()) grouped += MemberGroup(null, ungrouped)

        return grouped.filter { it.entries.isNotEmpty() }
    }
}
